PWGJE/TableProducer/emcalCorrectionTask.cxx

```diff
@@ -956,28 +1304,46 @@ struct EmcalCorrectionTask {
     trackGlobalIndex.reserve(nTracksInCol);
     fillTrackInfo<decltype(groupedTracks)>(groupedTracks, trackPhi, trackEta, trackGlobalIndex);
 
-    int nClusterInCol = mAnalysisClusters.size();
-    std::vector<float> clusterPhi;
-    std::vector<float> clusterEta;
-    clusterPhi.reserve(nClusterInCol);
-    clusterEta.reserve(nClusterInCol);
+    indexMapPair = matchTracksToCluster(mClusterPhi, mClusterEta, trackPhi, trackEta, maxMatchingDistance, 20);
+  }
 
-    // TODO one loop that could in principle be combined with the other
-    // loop to improve performance
-    for (const auto& cluster : mAnalysisClusters) {
-      // Determine the cluster eta, phi, correcting for the vertex
-      // position.
-      auto pos = cluster.getGlobalPosition();
-      clusterPhi.emplace_back(TVector2::Phi_0_2pi(pos.Phi()));
-      clusterEta.emplace_back(pos.Eta());
+  template <typename Collision>
+  void doSecondaryTrackMatching(Collision const& col, EMV0Legs const& v0legs, MatchResult& indexMapPair, std::vector<int64_t>& trackGlobalIndex, MyGlobTracks const& tracks)
+  {
+    auto groupedV0Legs = v0legs.sliceBy(perCollisionEMV0Legs, col.globalIndex());
+    int nLegsInCol = groupedV0Legs.size();
+    std::vector<float> trackPhi;
+    std::vector<float> trackEta;
+    // reserve memory to reduce on the fly memory allocation
+    trackPhi.reserve(nLegsInCol);
+    trackEta.reserve(nLegsInCol);
+    trackGlobalIndex.reserve(nLegsInCol);
+
+    float trackEtaEmcal = 0.f;
+    float trackPhiEmcal = 0.f;
+    // fillTrackInfo<decltype(groupedTracks)>(groupedTracks, trackPhi, trackEta, trackGlobalIndex);
+    for (const auto& leg : groupedV0Legs) {
+      auto track = tracks.iteratorAt(leg.trackId());
+      trackEtaEmcal = track.trackEtaEmcal();
+      trackPhiEmcal = track.trackPhiEmcal();
+      // Tracks that do not point to the EMCal/DCal/PHOS get default values of -999
+      // This way we can cut out tracks that do not point to the EMCal+DCal
+      if (trackEtaEmcal < TrackNotOnEMCal || trackPhiEmcal < TrackNotOnEMCal) {
+        continue;
+      }
+      if (trackMinPt > 0 && track.pt() < trackMinPt) {
+        continue;
+      }
+      trackPhi.emplace_back(RecoDecay::constrainAngle(trackPhiEmcal));
+      trackEta.emplace_back(trackEtaEmcal);
+      trackGlobalIndex.emplace_back(track.globalIndex());
     }
-    indexMapPair = matchTracksToCluster(clusterPhi, clusterEta, trackPhi, trackEta, maxMatchingDistance, 20);
+    indexMapPair = matchTracksToCluster(mClusterPhi, mClusterEta, trackPhi, trackEta, maxMatchingDistance, 20);
```

Comment:
consider using constexpr to avoid magic number:
```
constexpr int MAX_MATCHES_PER_CLUSTER = 20;  // Maximum number of tracks to match per cluster
indexMapPair = matchTracksToCluster(..., MAX_MATCHES_PER_CLUSTER);
```